🚀 2단계 - 로또(자동)

ljg294 opened the pull request:
* feature/lotteryDrawer-requirements
  * test codes based on requirements 
* feature/implement-lottery-application
  * implemented applications
  * test codes modified
* refactor/step1
  * refactored codes based on feedback
* refactor/lottery-application
  * refactored codes using stream()

neojjc2 commented:
#tag @neojjc2

neojjc2 commented on src/main/java/lottery/domain/Lotteries.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public int lotteriesTotalPrice(){
+        return this.lotteries.size() * 1000;
```
일단 `1000`은 매직넘버이니 추출되면 좋을 것 같습니다. (아래쪽에 1000과도 중복이니 제거가 가능할 것 같아요)
> https://hoonmaro.tistory.com/44

그리고 이 부분에 대한 검증도 추가되면 좋을 것 같습니다 😄 
구매금액과 관련된 객체가 하나 생성된다면 그쪽에 위임해보는것도 한 방법이 될 수 있을 것 같아요 🙇 
<img width="311" alt="image" src="https://github.com/next-step/java-lotto/assets/68527004/25ab5995-5c95-4112-840b-419315b88b8f">

neojjc2 commented on src/main/java/lottery/domain/Lotteries.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private static LotteryNumbers quickPick(){
+        List<Integer> allNumbers = IntStream.rangeClosed(1, 45).boxed().collect(Collectors.toList());
```
`1, 45, 6` 모두 매직넘버 이므로 정적상수로 추출되면 좋을 것 같습니다 🙇

neojjc2 replied:
`allNumbers`가 변경이 되는 값은 아니니 (1~45) 매번 `list`객체를 만드는 것보다 
정적 상수로 한번 초기화 해서 사용하는 것도 괜찮아 보입니다. (성능상)

neojjc2 commented on src/main/java/lottery/domain/vo/LotteryNumbers.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.stream.Collectors;
+
+public class LotteryNumbers {
+    private Set<Integer> numbers;
```
`Set`을 통한 중복제거 좋은데요?? 👍

neojjc2 commented on src/main/java/lottery/domain/vo/LotteryNumbers.java:
```diff
@@ -0,0 +1,46 @@
+    public int hashCode() {
+        return Objects.hash(numbers);
+    }
+}
```
다만 어딘가에서 이 부분에 대한 검증도 추가되면 좋을 것 같습니다 🙇 
<img width="269" alt="image" src="https://github.com/next-step/java-lotto/assets/68527004/50b09a8d-7114-4872-b136-c701ba560f17">

neojjc2 commented on src/main/java/lottery/view/InputView.java:
```diff
@@ -0,0 +1,19 @@
+
+public class InputView {
+
+    private static Scanner scan = new Scanner(System.in);
```
```suggestion
    private static final Scanner SCAN = new Scanner(System.in);
```

neojjc2 commented on src/test/java/lottery/service/LotteryDrawerTest.java:
```diff
@@ -0,0 +1,57 @@
+        BigDecimal expected = BigDecimal.valueOf(1.66);
+        assertThat(actual).isEqualTo(expected);
+    }
+}
```
테스트 작성도 전체적으로 좋습니다 💯

neojjc2 reviewed (changes requested):
안녕하세요 재규님 🙇 

2단계 구현 잘 진행해주셨습니다 😄 

조금 다듬어지면 좋을 것 같아서
소소한 의견들 몇 가지 드렸는데요,
한번 보시고 개선 검토 해주시면 감사하겠습니다 🙇 

그럼 재 리뷰요청 기다리고 있겠습니다 🙏

ljg294 commented:
* refactored codes based on feedback
    * validate input from system 
      * convert int to Integer -> able to test null and throw exception
    * used MagicNumber as Constant

ljg294 commented on src/main/java/lottery/domain/Lotteries.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.stream.LongStream;
+
+
+public class Lotteries {
```
네이밍 관련해서는 저도 고민이 있었습니다.
  웹 애플리케이션을 개발할때 Entity, Entities를 사용한 경우도 있었거든요.
  자세히 생각해보니 Lottery는 복권이고 우리가 하고자 하는것은 Lotto라서 그 복권게임을 할 수 있는 LottoTickets가 맞을 수도 있다고 생각하네요 (커밋후에 생각들었습니다. ㅎㅎ..)

ljg294 commented:
* refactored codes after lecture
    * renamed all objects lottery -> lottoTicket or else
    * test codes are also refactored
      * remove //Given //When // Then
    * LotteryNumbers removed, LottoNumber created
    
  개인 사정으로 인해 강의를 라이브로 못듣고 다시 들었습니다.
  강의 내용을 바탕으로 수정을 더 진행하였습니다.
  리팩토링하면서 naming도 다시 진행하였습니다.
  감사합니다 :)

neojjc2 commented on src/main/java/lottery/domain/LottoTicket.java:
```diff
@@ -0,0 +1,60 @@
+    public int hashCode() {
+        return Objects.hash(lottoNumbers);
+    }
+}
```
많은 부분 정리를 해주셨네요 👍

neojjc2 commented on src/main/java/lottery/domain/LottoTickets.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private static void validateMoney(Integer inputMoney){
+        if(Objects.isNull(inputMoney))
```
```suggestion
        if (Objects.isNull(inputMoney))
```

neojjc2 commented on src/main/java/lottery/domain/LottoTickets.java:
```diff
@@ -0,0 +1,67 @@
+    private static void validateMoney(Integer inputMoney){
+        if(Objects.isNull(inputMoney))
+            throw new IllegalArgumentException("구입 금액은 필수 입니다.");
+        if(inputMoney % LottoTicket.PRICE != 0)
```
```suggestion
        if (inputMoney % LottoTicket.PRICE != 0)
```

neojjc2 commented on src/main/java/lottery/domain/LottoTickets.java:
```diff
@@ -0,0 +1,67 @@
+            throw new IllegalArgumentException("구입 금액은 필수 입니다.");
+        if(inputMoney % LottoTicket.PRICE != 0)
+            throw new IllegalArgumentException("구입 금액은 1000 단위 입니다.");
+    }
```
검증 추가 좋습니다 💯 
`inputMoney`가 원시객체로 추출된다면 어떨까요??
적절하게 이런 역할들을 위임해볼 수 있을 것 같은데
이 부분은 다음단계 진행하시면서 한번 검토해주시면 좋을 것 같습니다 😄

neojjc2 commented on src/main/java/lottery/domain/vo/LottoNumber.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class LottoNumber {
```
💯 💯 💯 💯 💯

neojjc2 reviewed (approved):
안녕하세요 재규님 🙇 

요청사항 반영 깔끔하게 잘 진행해주셨습니다 💯 
진행하시느라 수고 많으셨습니다 😄 

그럼 이 구조에서
다음단계 요구사항들이 어떻게 반영될지 궁금하네요 😄 

이만 `merge`하겠습니다.
수고하셨습니다 🙇
